src/plugins/tracking-mark.tsx

```diff
@@ -0,0 +1,120 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { getMarkStyle, TrackAttrsButton } from '@manuscripts/style-guide'
+import {
+  NodeAttrChange,
+  NodeChange,
+  trackChangesPluginKey,
+} from '@manuscripts/track-changes-plugin'
+import { schema } from '@manuscripts/transform'
+import { EditorState, Plugin } from 'prosemirror-state'
+import { Decoration, DecorationSet } from 'prosemirror-view'
+import React from 'react'
+
+type RenderReactComponent = (
+  child: React.ReactElement,
+  container: HTMLElement
+) => void
+
+type TrackingMarkProps = {
+  renderReactComponent: RenderReactComponent
+}
+
+/**
+ * This plugin adds tracking marks for PM node that has just attributes without content
+ * Like: (bibliography_item, citation, equation, inline_equation, figure)
+ */
+export default (props: TrackingMarkProps) =>
+  new Plugin<DecorationSet>({
+    state: {
+      init: (tr, state) => {
+        return DecorationSet.create(state.doc, buildDecorations(state, props))
+      },
+      apply: (tr, decorationSet, oldState, newState) => {
+        const meta = tr.getMeta('track-changes-set-change-statuses')
+
+        if (meta) {
+          return DecorationSet.create(
+            newState.doc,
+            buildDecorations(newState, props)
+          )
+        }
+
+        return decorationSet
+      },
+    },
+    props: {
+      decorations: (state) => {
+        return DecorationSet.create(state.doc, buildDecorations(state, props))
+      },
+    },
+  })
+
+const buildDecorations = (state: EditorState, props: TrackingMarkProps) => {
+  const changeSet = trackChangesPluginKey.getState(state)?.changeSet
+
+  if (!changeSet) {
+    return []
+  }
+
+  const metaNodeChanges = [
+    ...(changeSet.nodeChanges as NodeChange[]),
+    ...(changeSet.nodeAttrChanges as NodeAttrChange[]),
+  ].filter(({ nodeType }) => nodeType === schema.nodes.bibliography_item.name)
+
+  const decorations: Decoration[] = []
+
+  metaNodeChanges.map((change) => {
+    const { style, className, showAttrsPopper } = getMarkStyle(
+      change.dataTracked
+    )
+
+    decorations.push(
+      Decoration.node(
+        change.from,
+        change.to,
+        {
+          class: className,
+          style: `background: ${style.background}; text-decoration: ${style.textDecoration}; display: ${style.display}; position: relative;`,
+        },
+        { id: change.id }
+      )
+    )
+
+    if (change.type === 'node-attr-change' && showAttrsPopper) {
+      decorations.push(
+        Decoration.widget(
+          change.from + 1,
+          addAttrsTrackingButton(change, props),
+          { key: change.id }
+        )
+      )
+    }
+  })
+
+  return decorations
+}
+
+const addAttrsTrackingButton = (
```

Comment:
It seems to me that the plugin is hyper aware of what needs to happen. Also the responsibilities of rendering things is generally bestowed upon the views and plugins mostly just manipulate data and transactions. The plugin shouldn't be aware of rendering methods, react components and html elements. be creating DOM. So at the very least plugin should receive a function that will completely take care of all the rendering hassle and here it will look only like these: addTrackingButton(change.id). So that way plugin will not be aware of renderReactComponent and will not need any imports from styleguide. It will just get a function from props, that will receive a change id as an argument and will return a dom node.

Reply:
updated